signer/services.go

```diff
@@ -1,17 +1,83 @@
 package signer
 
 import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
 	"sync"
+	"syscall"
 
 	tmLog "github.com/tendermint/tendermint/libs/log"
 	tmOS "github.com/tendermint/tendermint/libs/os"
 	tmService "github.com/tendermint/tendermint/libs/service"
 )
 
-func WaitAndTerminate(logger tmLog.Logger, services []tmService.Service) {
+func RequireNotRunning(pidFilePath string) error {
+	if _, err := os.Stat(pidFilePath); err != nil {
+		if os.IsNotExist(err) {
+			// lock file does not exist, can continue starting daemon
+			// or performing other tasks that require horcrux daemon to be stopped.
+			return nil
+		}
+		return fmt.Errorf("unexpected error while checking for existence of lock file at %s: %w", pidFilePath, err)
+	}
+
+	lockFile, err := os.ReadFile(pidFilePath)
+	if err != nil {
+		return fmt.Errorf("error reading lock file: %s, %w", pidFilePath, err)
+	}
+
+	pid, err := strconv.ParseInt(strings.TrimSpace(string(lockFile)), 10, 64)
+	if err != nil {
+		return fmt.Errorf("unexpected error parsing PID from lock file: %s. manual deletion of PID file required. %w",
+			pidFilePath, err)
+	}
+
+	process, err := os.FindProcess(int(pid))
+	if err != nil {
+		return fmt.Errorf(`unclean shutdown detected. PID file exists at %s but PID %d can not be found.
+manual deletion of PID file required. %w`, pidFilePath, pid, err)
+	}
+
+	err = process.Signal(syscall.Signal(0))
+	if err == nil {
+		return fmt.Errorf("horcrux is already running on PID: %d", pid)
+	}
+	if err.Error() == "os: process already finished" {
+		return fmt.Errorf(`unclean shutdown detected. PID file exists at %s but PID %d is not running.
+manual deletion of PID file required`, pidFilePath, pid)
+	}
+
+	errno, ok := err.(syscall.Errno)
+	if !ok {
+		return fmt.Errorf("unexpected error type from signaling horcrux PID: %d", pid)
+	}
+	switch errno {
+	case syscall.ESRCH:
+		return fmt.Errorf("search error while signaling horcrux PID: %d", pid)
+	case syscall.EPERM:
+		return fmt.Errorf("permission denied accessing horcrux PID: %d", pid)
+	}
+	return fmt.Errorf("unexpected error while signaling horcrux PID: %d", pid)
+}
+
+func WaitAndTerminate(logger tmLog.Logger, services []tmService.Service, pidFilePath string) {
 	wg := sync.WaitGroup{}
 	wg.Add(1)
+
+	err := os.WriteFile(
+		pidFilePath,
+		[]byte(fmt.Sprintf("%d\n", os.Getpid())),
+		0600,
+	)
```

Comment:
What happens if two horcrux processes are started at the same time, neither sees the pid file, and both attempt to write their pid?

This probably needs to switch to `os.OpenFile(pidFilePath, os.O_WRONLY | os.O_CREATE | os.O_EXCL, 0600)` to ensure only one process can create it. And it would be good to get a unit test that uses several goroutines running this function, ensuring that only one succeeds.